File: app14_LUA_UNITY/backend/spec/services/reference_validator_spec.lua
require("spec.spec_helper")
local ReferenceValidator = require("app.services.reference_validator")

describe("ReferenceValidator", function()
  local validator = ReferenceValidator.new(SEEDS_ROOT)

  it("accepts a known OWASP ref", function()
    assert.has_no.errors(function() validator:assert_owasp_refs_valid({ "A03:2021" }, "TEST") end)
  end)

  it("accepts a known MITRE ATLAS ref", function()
    assert.has_no.errors(function() validator:assert_mitre_refs_valid({ "AML.T0051" }, "TEST") end)
  end)

  it("rejects an unknown OWASP ref", function()
    local ok, err = pcall(function() validator:assert_owasp_refs_valid({ "A99:2099" }, "TEST") end)
    assert.is_false(ok)
    assert.are.equal("A99:2099", err.value)
    assert.are.equal("owasp_refs", err.field)
    assert.are.equal("TEST", err.card_id)
  end)

  it("rejects an unknown MITRE ref", function()
    assert.has_error(function() validator:assert_mitre_refs_valid({ "AML.T9999" }, "TEST") end)
  end)

  it("never raises for an empty refs list", function()
    assert.has_no.errors(function() validator:assert_owasp_refs_valid({}, "TEST") end)
    assert.has_no.errors(function() validator:assert_mitre_refs_valid({}, "TEST") end)
  end)
end)

File: app13_ruby_FastApi/backend/spec/services/reference_validator_spec.rb
# frozen_string_literal: true

require_relative "../spec_helper"

RSpec.describe ReferenceValidator do
  subject(:validator) { described_class.new(seeds_root: SEEDS_ROOT) }

  it "accepts a known OWASP ref" do
    expect { validator.assert_owasp_refs_valid!(["A03:2021"], "TEST") }.not_to raise_error
  end

  it "accepts a known MITRE ATLAS ref" do
    expect { validator.assert_mitre_refs_valid!(["AML.T0051"], "TEST") }.not_to raise_error
  end

  it "rejects an unknown OWASP ref" do
    expect { validator.assert_owasp_refs_valid!(["A99:2099"], "TEST") }
      .to raise_error(CardDecodeError::UnknownReference) do |error|
        expect(error.value).to eq("A99:2099")
        expect(error.field).to eq("owasp_refs")
        expect(error.card_id).to eq("TEST")
      end
  end

  it "rejects an unknown MITRE ref" do
    expect { validator.assert_mitre_refs_valid!(["AML.T9999"], "TEST") }
      .to raise_error(CardDecodeError::UnknownReference)
  end

  it "never raises for an empty refs list" do
    expect { validator.assert_owasp_refs_valid!([], "TEST") }.not_to raise_error
    expect { validator.assert_mitre_refs_valid!([], "TEST") }.not_to raise_error
  end
end
